common/src/bitset.rs

```diff
@@ -196,6 +196,30 @@ impl BitSet {
         Ok(())
     }
 
+    pub fn deserialize(mut reader: impl Read) -> io::Result<Self> {
+        let mut max_buf = vec![0; 4];
+        reader.read_exact(&mut max_buf)?;
+        let max_value =
+            u32::from_le_bytes(max_buf.try_into().expect("Max value should be 4 bytes"));
+        let mut tinysets_buf = Vec::new();
+        reader.read_to_end(&mut tinysets_buf)?;
+        let mut tinysets = Vec::new();
+        let mut len = 0;
+        for chunk in tinysets_buf.chunks_exact(8) {
+            let slice: &[u8; 8] = chunk.try_into().unwrap();
+            let tinyset = TinySet::deserialize(slice.clone());
+            len += tinyset.len() as u64;
+            tinysets.push(tinyset);
+        }
```

Comment:
you could avoid a few allocations by (1) presizing the buffers and (2) parsing the tinysets as we read the file:
```rust
let buckets = num_buckets(max_value);
let mut tinysets = Vec::with_capacity(buckets);

for _ in 0..buckets {
    let mut buf = [0; 8];
    reader.read_exact(&mut buf)?;
    let tinyset = TinySet::deserialize(buf);
    tinysets.push(tinyset);
}

Reply:
sweet! thanks for these optimization suggestions.